### PR TITLE
OCPNODE-1717: Update config node spec while explicitly updating the cgroups mode

### PR DESCRIPTION
**Pull request by sairameshv**
- Future OCP releases will default to cgroupsv2
- During the upgrades, in order to honor the cgroupsv1 on the existing cluster, there should be a reference in the node spec for the future code to depend on deciding the cgroup version.
- Hence, this code updates the config node spec whenever the cgroupsv1 is explicitly set

### CODE REVIEW

**Comment by sairameshv** [comment]
/hold cancel
This PR has been tested for upgrade to the current code base, ready for the review and hence cancelling the hold

File: pkg/controller/kubelet-config/kubelet_config_nodes.go
Comment (yuqi-zhang) [inline pkg/controller/kubelet-config/kubelet_config_nodes.go:84]: nit: technically this code is only in 4.13. Maybe change this comment to:

`This code only exists in 4.13 to set the default to v1, if the user did not otherwise provide it. All clusters must upgrade through this 4.13.z before upgrade to 4.14 to preserve v1 defaults`

Something along those lines?

File: pkg/controller/kubelet-config/kubelet_config_nodes.go
Comment (yuqi-zhang) [inline pkg/controller/kubelet-config/kubelet_config_nodes.go:279]: Does this work as we intend? As I understand it, this will only generate the machineconfig with the correct config, but doesn't actually create the nodeConfig object to cluster.

This combined with the above `ctrl.configClient.ConfigV1().Nodes().Update(context.TODO(), nodeConfig, metav1.UpdateOptions{})`, is it possible that we would run into some race condition where the MCO generates the machineconfigs with an empty nodes.config before the above sync has a chance to run?

Or, hmm, maybe I am overthinking it. Just wanted to make sure

**Review by yuqi-zhang — COMMENTED** [review]
I think this path should work, but we would need to be a bit careful of how we update the nodes config.

So the expectation here is for upgraded clusters, the user would need to manually set the cgroups to v2 via nodes.config, **or** delete the nodes.config object in cluster (thereby defaulting it back to the RHCOS default (of v2)?

We should definitely document the behaviour change and how users should do this carefully. Let's also talk to the release team about edge blocking, just so they are ok with this solution.

File: pkg/controller/kubelet-config/kubelet_config_nodes.go
Comment (sairameshv) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:84]: Sure, Thanks !!

File: pkg/controller/kubelet-config/kubelet_config_nodes.go
Comment (sairameshv) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:279]: > Does this work as we intend? As I understand it, this will only generate the machineconfig with the correct config, but doesn't actually create the nodeConfig object to cluster.

The nodeConfig object is by default created, present in the cluster.

> This combined with the above ctrl.configClient.ConfigV1().Nodes().Update(context.TODO(), nodeConfig, metav1.UpdateOptions{}), is it possible that we would run into some race condition where the MCO generates the machineconfigs with an empty nodes.config before the above sync has a chance to run?

Yeah, our goal here is to mainly update the config node object with the `cgroupv1` context. Creation of the respective machineconfig will only add the `kernelArgs` to enable `cgroupsv1` which is already taken care as part of the base templates (explicit setting of cgroupsv1). So, we need not worr about running into any race condition.

Comment (yuqi-zhang) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:282]: > The nodeConfig object is by default created, present in the cluster.

Right, the default used to be empty.

> Yeah, our goal here is to mainly update the config node object with the cgroupv1 context. Creation of the respective machineconfig will only add the kernelArgs to enable cgroupsv1 which is already taken care as part of the base templates (explicit setting of cgroupsv1). So, we need not worr about running into any race condition.

Ah, so you verified that the base kernalArgs and the kernalArgs generated by setting the nodes.config object is the **exact same** ?

I was concerned about our karg merging logic since I don't remember if it was correct or not. If you can verify that it is exact same and no extra rendered config needs to be generated, then I think we are good to go

Comment (sairameshv) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:282]: Yes, the kArgs are the same but there would be a new rendered MC as there is a `97-{master/worker}-generated-kubelet` are created.

**Review by yuqi-zhang — APPROVED** [review]
Ok, based on the CI run I think the behaviours are as expected. We end up with duplicated kargs for now, but that shouldn't be an issue once we upgrade to 4.14.

**Comment by rphillips** [comment]
/hold

@haircommander brought up a great point... Does this PR upgrade successfully to? https://github.com/openshift/machine-config-operator/pull/3782

**Comment by yuqi-zhang** [comment]
Since this is a bit of an exception to the regular process (4.13 only) we may need to either:
1. ask for someone to help merge this PR by adding the valid bug label, or
2. create a dummy 4.14 bug, mark as verified directly, then doing a /jira cherry-pick on here

**Comment by sairameshv** [comment]
> > Does this PR upgrade successfully to? #3782
> 
> In theory it should, the MCO upon upgrade will generate a new config without the kargs in the base, but since the kargs still exist in the update node.configs object generated 97-generated-kubeletconfig, it will end up in the render and just be a no-op change by itself. @sairameshv did you verify that path? I seem to recall you did

Yes, this upgrade has been tested.

**Comment by rphillips** [comment]
/hold cancel

ok. thanks!

**Pull request merged**
